File: Erettsegi/emelt/2015_may_eng/Tanciskola.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Prog{
    class Tanciskola{
        static void Main(string[] args){
            string[] file = File.ReadAllLines("tancrend.txt");
		    List<Tanc> tancok = new List<Tanc>();
		
		    for(int k = 0; k < file.Length; k += 3) {
			    tancok.Add(new Tanc(file[k], file[k + 1], file[k + 2]));
		    }
		
		    Console.WriteLine("Első tánc neve: " + tancok[0].category + ", az utolsóé: " + tancok[tancok.Count - 1].category);
		
		    int szambaCounter = 0;
		    foreach(Tanc dansz in tancok) {
			    if(dansz.category.Equals("samba")) {
				    ++szambaCounter;
			    }
		    }
		    Console.WriteLine("Összesen " + szambaCounter + "-an szambásztak");
		
		    List<String> viliCat = new List<String>();
		    foreach(Tanc dansz in tancok) {
			    if(dansz.woman.Equals("Vilma") && !viliCat.Contains(dansz.category)) {
				    viliCat.Add(dansz.category);
			    }
		    }
		    Console.WriteLine("Vilma által táncolt kategóriák: " + viliCat.Aggregate((k, l) => k + ", " + l));
		    Console.WriteLine("Írj be 1 kategóriát!");
		
			String readCat = Console.ReadLine();
			bool hasPrinted = false;
			foreach(Tanc dansz in tancok) {
			    if(dansz.woman.Equals("Vilma") && dansz.category.Equals(readCat)) {
				    Console.WriteLine("Vilma a " + readCat + " kategóriában " + dansz.man + "-val táncolt");
				    hasPrinted = true;
				}
			 }
			 if(!hasPrinted) {
				 Console.WriteLine("Vilma a " + readCat + " kategóriában nem táncolt");
			 }
		    
		    HashSet<String> csajok = new HashSet<String>();
		    HashSet<String> skacok = new HashSet<String>();
		    List<Szereplo> lanyok = new List<Szereplo>();
		    List<Szereplo> fiuk = new List<Szereplo>();

		    StreamWriter output = new StreamWriter("szereplok.txt");
		    foreach(Tanc dans in tancok) {
			    csajok.Add(dans.woman);
			    skacok.Add(dans.man);
		    }
		    foreach(String cs in csajok) {
			    lanyok.Add(new Szereplo(cs));
		    }
		    foreach(String fik in skacok) {
			    fiuk.Add(new Szereplo(fik));
		    }
		    output.Write("Lányok: ");
		
		    for(int k = 0; k < lanyok.Count; ++k) {
			    if(k != lanyok.Count - 1) {
				    output.Write(lanyok[k] + ", ");
			    }else{
				    output.Write(lanyok[k]);
			    }
		    }
		    output.WriteLine();
		    output.Write("Fiúk: ");
		    for(int k = 0; k < fiuk.Count; ++k) {
			    if(k != fiuk.Count - 1) {
				    output.Write(fiuk[k] + ", ");
			    }else{
				    output.Write(fiuk[k]);
			    }
		    }
		    output.Close();
		
		    foreach(Tanc dan in tancok) {
			    foreach(Szereplo csaj in lanyok) {
				    if(dan.woman.Equals(csaj.name)) {
					    ++csaj.alkalmak;
				    }
			    }
			    foreach(Szereplo fiu in fiuk) {
				    if(dan.man.Equals(fiu.name)) {
					    ++fiu.alkalmak;
				    }
			    }
		    }

            lanyok = lanyok.OrderByDescending(k => k.alkalmak).ToList();
            fiuk = fiuk.OrderByDescending(k => k.alkalmak).ToList();
		
		    int lanyMax = lanyok[0].alkalmak;
		    int fiuMax = fiuk[0].alkalmak;
		
		    Console.Write("A legtöbbet táncolt lányok: ");
		    foreach(Szereplo la in lanyok) {
			    if(la.alkalmak == lanyMax) {
				    Console.Write(la.name + " ");
			    }
		    }
		    Console.WriteLine();
		    Console.Write("A legtöbbet táncolt fiúk: ");
		    foreach(Szereplo fi in fiuk) {
			    if(fi.alkalmak == fiuMax) {
				    Console.Write(fi.name + " ");
			    }
		    }
            Console.Read();
	    }
	    class Tanc{
		    public String category, woman, man;
		    public Tanc(String cat, String wom, String ma) {
			    category = cat;
			    woman = wom;
			    man = ma;
		    }
	    }
	    class Szereplo{
		    public String name;
		    public int alkalmak = 0;
		    public Szereplo(String neim) {
			    name = neim;
		    }
		    public override string ToString() {
			    return name;
		    }
		    public int getAlkalmak() {
			    return alkalmak;
		    }
	    }
    }
}

File: erettsegi/emelt/2015_may_eng/java/Tanciskola.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Tanciskola {

    public static void main(String[] args) throws IOException {
        var lines = Files.readAllLines(Path.of("tancrend.txt"));
        var tancok = new ArrayList<Tanc>();

        for(var i = 0; i < lines.size(); i += 3) {
            tancok.add(new Tanc(lines, i));
        }

        System.out.println("Első tánc neve: " + tancok.get(0).category + ", az utolsóé: " + tancok.get(tancok.size() - 1).category);

        var szambaCounter = 0;
        for(var dansz : tancok) {
            if(dansz.category.equals("samba")) {
                ++szambaCounter;
            }
        }

        System.out.println("Összesen " + szambaCounter + "-an szambásztak");

        var viliCat = new ArrayList<String>();
        for(var dansz : tancok) {
            if(dansz.woman.equals("Vilma") && !viliCat.contains(dansz.category)) {
                viliCat.add(dansz.category);
            }
        }

        System.out.println("Vilma által táncolt kategóriák: " + String.join(", ", viliCat));

        var readCat = System.console().readLine("Írj be 1 kategóriát: ");
        var hasPrinted = false;

        for(var dansz : tancok) {
            if(dansz.woman.equals("Vilma") && dansz.category.equals(readCat)) {
                System.out.println("Vilma a " + readCat + " kategóriában " + dansz.man + "-val táncolt");
                hasPrinted = true;
            }
        }

        if(!hasPrinted) {
            System.out.println("Vilma a " + readCat + " kategóriában nem táncolt");
        }

        var lanyokToTancalkalmak = new HashMap<String, Integer>();
        var fiukToTancalkalmak = new HashMap<String, Integer>();

        for(var dans : tancok) {
            lanyokToTancalkalmak.put(dans.woman, lanyokToTancalkalmak.getOrDefault(dans.woman, 0) + 1);
            fiukToTancalkalmak.put(dans.man, fiukToTancalkalmak.getOrDefault(dans.man, 0) + 1);
        }

        Files.writeString(Path.of("szereplok.txt"), "Lányok: " + String.join(", ", lanyokToTancalkalmak.keySet()) +
                                                    "\nFiúk: " + String.join(", ", fiukToTancalkalmak.keySet()));
        var grillMax = 0;
        for(var v : lanyokToTancalkalmak.values()) {
            if(v > grillMax) {
                grillMax = v;
            }
        }

        System.out.print("A legtöbbet táncolt lányok: ");
        for(var e : lanyokToTancalkalmak.entrySet()) {
            if(e.getValue() == grillMax) {
                System.out.print(e.getKey() + ' ');
            }
        }

        var boiMax = 0;
        for(var v : fiukToTancalkalmak.values()) {
            if(v > boiMax) {
                boiMax = v;
            }
        }

        System.out.print("\nA legtöbbet táncolt fiúk: ");
        for(var e : fiukToTancalkalmak.entrySet()) {
            if(e.getValue() == boiMax) {
                System.out.print(e.getKey() + ' ');
            }
        }
    }
}